src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -75,6 +75,80 @@ typedef struct WorkerAggregateWalkerContext
 } WorkerAggregateWalkerContext;
 
 
+/*
+ * QueryTargetList encapsulates the necessary fields to form
+ * worker query's target list.
+ */
+typedef struct QueryTargetList
+{
+	List **targetEntryList; /* pointer to the list of target entries */
+	AttrNumber *targetProjectionNumber; /* pointer to the index of the last entry */
+} QueryTargetList;
+
+
+/*
+ * QueryGroupClause encapsulates the necessary fields to form
+ * worker query's group by clause.
+ */
+typedef struct QueryGroupClause
+{
+	List **groupClauseList; /* pointer to the list of group clause entries */
+	Index *nextSortGroupRefIndex; /* pointer to the index of the largest sort group reference index */
+} QueryGroupClause;
+
+
+/*
+ * QueryDistinctClause encapsulates the necessary fields to form
+ * worker query's DISTINCT/DISTINCT ON parts.
+ */
+typedef struct QueryDistinctClause
+{
+	List **workerDistinctClause; /* pointer to the list of distinct clause entries */
+	bool *workerHasDistinctOn;
+} QueryDistinctClause;
+
+
+/*
+ * QueryWindowClause encapsulates the necessary fields to form
+ * worker query's window clause.
+ */
+typedef struct QueryWindowClause
+{
+	List **workerWindowClauseList;
+	bool *hasWindowFunctions;
+	Index *nextSortGroupRefIndex;
```

Comment:
Why are these pointers? Can't we just modify the values if we have a pointer to the struct?

Reply:
I think we could. But, my idea was to make sure that we have a singleton `workerExtendedOpNode ` and all the fields in these structs directly point to the `workerExtendedOpNode`'s fields to prevent any weird bugs in the future.

I was mostly suspicious about `nextSortGroupRefIndex ` which is shared by multiple structs.

Do you think would it be better to not have pointers here?

Reply:
I think using a lot of pointers is prone to bugs such as segmentation faults being introduced. I like having pointers to structs that are allocated on the stack (as you did), and then modifying the fields.

Reply:
But in this case, they are all references to some already allocated variables:

```C
Index nextSortGroupRefIndex = 0;

... 
queryGroupClause.nextSortGroupRefIndex = &nextSortGroupRefIndex;
``` 

One minor reason that I try to push back on doing that is we'd end up with assigning the values back to some other structs in between function calls such as:

```C

/* process each part of the query in order to generate the worker query's parts */
ProcessTargetListForWorkerQuery(originalTargetEntryList, extendedOpNodeProperties,
									&queryTargetList, &queryGroupClause);

...

queryWindowClause.nextSortGroupRefIndex = queryGroupClause.nextSortGroupRefIndex;
ProcessWindowFunctionsForWorkerQuery(originalWindowClause, originalTargetEntryList,
										 &queryWindowClause, &queryTargetList);

```
Becuase those are acutally global variables within the function.

And, missing such things could lead to bugs that are hard to understand, whereas segfault are pretty simple to recognise 😃 

But, I'm OK to change if you still think that non-pointer fields are better.

Reply:
(I can see that using these structs on other places would prone to what you've described, but, I don't think these are common structs to use at any other place?)

Reply:
> One minor reason that I try to push back on doing that is we'd end up with assigning the values back to some other structs in between function calls such as:

That could be a good reason to keep a pointer in case of `nextSortGroupRefIndex`, but then I again wonder whether the separation really makes sense and the functions shouldn't just be working on the `MultiExtendedOp` directly.

Anyway, I think de-pointering everything that isn't shared between structs would be acceptable.

> And, missing such things could lead to bugs that are hard to understand, whereas segfault are pretty simple to recognise 😃

and extremely annoying to customers